### ISSUE TITLE
False positive of `DMI_RANDOM_USED_ONLY_ONCE`

### ISSUE THREAD

**Issue by seyfahni** [spotbugs/spotbugs#1539]
Updating from version `4.1.2` to `4.2.3` resulted in some findings of `DMI_RANDOM_USED_ONLY_ONCE`, but fixing them by creating a class field for the random instance did not work.
Update: The issue was introduced with `4.2.3`. We upgraded to `4.2.2` and there are no false positives.

## False positive cases
I created a pull request that reproduces the issue: #1540
It contains the following cases:
- `Random` via method argument:
  ```java
  public class Issue1539Argument {
      public static boolean shouldNotReport(final Random random) {
          return random.nextBoolean();
      }
  }
  ```
- `Random` provided during instance creation:
  ```java
  public class Issue1539Instance {
      private final Random random;
  
      public Issue1539Instance(long seed) {
          random = new Random(seed);
      }
  
      public boolean shouldNotReport() {
          return random.nextBoolean();
      }
  }
  ```
- `static final Random` class constant:
  ```java
  public class Issue1539Static {
      private static final Random RANDOM = new Random();
  
      public boolean shouldNotReport() {
          return RANDOM.nextBoolean();
      }
  }
  ```
- `Random` from a provider method:
  ```java
  public class Issue1539ThreadLocal {
      public boolean shouldNotReport() {
          Random random = ThreadLocalRandom.current();
          return random.nextBoolean();
      }
  }
  ```

Please only report new kinds of false positives.

## Example

Example code that is correctly reported:
```java
public class RandomCondition {

    private final int success;
    private final int total;

    public RandomCondition(int success, int total) {
        this.success = success;
        this.total = total;
    }

    public boolean isMet() {
        return new Random().nextInt(total) < success;
    }
}
```

My fix that I thought should not be reported, but is:
```java
public class RandomCondition {

    private final Random random = new Random();
    private final int success;
    private final int total;

    public RandomCondition(int success, int total) {
        this.success = success;
        this.total = total;
    }

    public boolean isMet() {
        return random.nextInt(total) < success;
    }
}
```

## maven project configuration
```xml
  <dependencies>
    <dependency>
      <groupId>com.github.spotbugs</groupId>
      <artifactId>spotbugs-annotations</artifactId>
      <version>4.2.3</version>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>com.github.spotbugs</groupId>
        <artifactId>spotbugs-maven-plugin</artifactId>
        <executions>
          <execution>
            <goals>
              <goal>check</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
    </plugins>

    <pluginManagement>
      <plugins>
        <plugin>
          <groupId>com.github.spotbugs</groupId>
          <artifactId>spotbugs-maven-plugin</artifactId>
          <version>4.2.3</version>
          <configuration>
            <effort>max</effort>
            <threshold>low</threshold>
          </configuration>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>

  <reporting>
    <plugins>
      <plugin>
        <groupId>com.github.spotbugs</groupId>
        <artifactId>spotbugs-maven-plugin</artifactId>
      </plugin>
    </plugins>
  </reporting>
```

**Comment by ThrawnCA** [comment]
Perhaps the trouble is that you have created an instance field, not a class field? Which means you'll still be unnecessarily constructing `Random` objects, each time you instantiate a `RandomCondition`.

**Comment by seyfahni** [comment]
`static` also fails.

And on that note, sometimes you need an instance field for Random, especially if you want to implement seeded logic like world generation.

I adjusted the test cases accordingly.

**Comment by seyfahni** [comment]
Even taking a Random instance as argument and only using it once will trigger `DMI_RANDOM_USED_ONLY_ONCE`.
```java
void useRandom(Random random) {
    random.nextInt();
}
```

I also added a test case,

**Comment by dodgex** [comment]
I can confirm that this issue happens! We have a `static final` field where the issue is reported.

Luckily we are able to use `@SuppressFBWarnings` until it is fixed.

Spotbugs 4.1.3 (invoked during build by Spotbugs Maven Plugin 4.0.4)

**Comment by inponomarev** [comment]
We also came across this issue in the following fragment of code:

```java
 Random r = ThreadLocalRandom.current();
 String projectName = String.format("Project%d", r.nextInt());
```

**Comment by seyfahni** [comment]
> ```java
>  Random r = ThreadLocalRandom.current();
>  String projectName = String.format("Project%d", r.nextInt());
> ```

I've added this to / modified the existing tests to also check for that one, previously I had two tests (`Argument` and `ThreadLocal`) that did actually only test the "Random as method argument used exactly once" false positive.

**Comment by alabotski** [comment]
Version 4.2.3
```
    private static final Random RANDOM = new Random();
...
    public boolean areUserMessagesExist(String login) {
        return RANDOM.nextBoolean();
    }
```
And I have error - DMI_RANDOM_USED_ONLY_ONCE

On version 4.2.2 all works fine

**Comment by mweber03** [comment]
I also see this issue when going from 4.2.0 to 4.2.3

**Comment by terheyden** [comment]
Same, in `4.2.3` this check is now throwing even when using a `private static final Random RAND = new SecureRandom();`

**Comment by matthewlowry** [comment]
Just to join the chorus:I too have observed this issue. For me it was upgrading from 3.x to 4.x - but experience as described by others here.

The pattern is triggering on legitimate violations but also triggering on code that conforms to the good practice the pattern aims to enforce.

For us the work-around is simply configure the Maven plugin with an exclude filter file and list that:
`pom.xml:`
```
                <configuration>
                    <excludeFilterFile>spotbugs-excludebugs.xml</excludeFilterFile>
                </configuration>
```
`spotbugs-excludebugs.xml`:
```
<FindBugsFilter>
    <Match>
        <Bug pattern="DMI_RANDOM_USED_ONLY_ONCE, ..." />
    </Match>
</FindBugsFilter>
```

This will make it a bit easier down the road to remove the exclusion once this issue is fixed (compared to say using `@SuppressFBWarnings` on locations where a false positive is triggering)
